File: pa1/src/Lex.java
// Lex.java.java - Sorts an input file by unicode order
// Chris Griffis <chdgriff>

import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;

public class Lex {
  public static void main(String[] args) throws FileNotFoundException, IOException {
    if (args.length != 2) {
      System.err.println("Incorrect usage Lex.java: Lex [file_input] [file_output]");
    }
    Scanner in = new Scanner(new File(args[0]));
    String input = in.useDelimiter("\\Z").next();
    in.close();
    String[] inArr= input.split("\\n");
    
    List list = new List();
    
    boolean inserted = false;
    for (int i = 0; i < inArr.length; i++) {
      inserted = false;
      while (!inserted) {
        if (list.length() == 0) { // If empty list
          list.prepend(i);
          inserted = true;
          list.moveFront();
        }
        else if (inArr[i].compareTo(inArr[list.get()]) < 0) { // if less than cursor
          if (list.index() == 0) { // if cursor is at beginning
            list.prepend(i);
            inserted = true;
          }
          else {
            list.movePrev();
            if (inArr[i].compareTo(inArr[list.get()]) > 0) { // Greater than new cursor position
              list.insertAfter(i);
              inserted = true;
            }
          }
        }
        else if (inArr[i].compareTo(inArr[list.get()]) > 0) { // if greater than cursor
          if (list.index() == list.length()-1) { // if cursor is at the end
            list.append(i);
            inserted = true;
          }
          else {
            list.moveNext();
            if (inArr[i].compareTo(inArr[list.get()]) < 0) { // Not at end and less than cursor
              list.insertBefore(i);
              inserted = true;
            }
          }
        }
        else {
          list.insertBefore(i);
          inserted = true;
        }
      }
    }
    
    FileWriter output = new FileWriter(args[1]);
    
    list.moveFront();
    while (list.index() != -1) {
      if (list.index() != 0) output.write('\n'); 
      output.write(inArr[list.get()]);
      list.moveNext();
    }
    output.close();
  }
}

File: pa2/src/Lex.c
// Lex.c - Takes lines from a command line file and outputs a file sorted
// Chris Griffis <chdgriff>
// pa2

#include "List.h"
#include <stdlib.h>
#include <string.h>

#define FILE_LEN 100
#define CHAR_LEN 60

int main(int argc, char* argv[]) {
  FILE *in, *out;
  char buffer[CHAR_LEN];
  char **lines;
  int lineCount;

  List L;
  int inserted, fileLength;

  if (argc != 3) {
    printf("Usage Lex.c: %s <input file> <output file>\n", argv[0]);
    exit(1);
  }

  in = fopen(argv[1], "r");
  out = fopen(argv[2], "w");
  if ( in == NULL) {
    printf("Lex.c Error: unable to open %s for reading\n", argv[1]);
    exit(1);
  }
  if (out == NULL) {
    printf("Lex.c Error: unable to open file %s for writing\n", argv[2]);
    exit(1);
  }


  fileLength = FILE_LEN;
  lines = malloc(fileLength * sizeof(char *)); // mallocs array for pointers to each line
  lineCount = 0;
  while (fgets(buffer, CHAR_LEN, in) != NULL) {
    if (lineCount == fileLength) { // if array is full realloc by 2 times size
      fileLength *= 2;
      lines = realloc(lines, fileLength * sizeof(char *));
    }
    lines[lineCount] = malloc(CHAR_LEN * sizeof(char)); // Mallocs space for line
    strcpy(lines[lineCount++], buffer); // copies string into array
  }

  L = newList();
  for (int i = 0; i < lineCount; i++) { // Loops through each line in array
    inserted = 0;
    while (inserted == 0) {
      if (length(L) == 0) { // if empty List
        prepend(L, 0);
        inserted = 1;
        moveFront(L);
      }
      else if (strcmp(lines[i], lines[get(L)]) < 0) { // If less than cursor
        if (index(L) == 0) { // If cursor is at front
          prepend(L, i);
          inserted = 1;
        }
        else { // If cursor is in the middle
          movePrev(L);
          if (strcmp(lines[i], lines[get(L)]) > 0) { // If greater than new cursor
            insertAfter(L, i);
            inserted = 1;
          }
        }
      }
      else if (strcmp(lines[i], lines[get(L)]) > 0) { // If greater than cursor
        if (index(L) == length(L)-1) { // If cursor is at back
          append(L, i);
          inserted = 1;
        }
        else { // If greater than new cursor
          moveNext(L);
          if (strcmp(lines[i], lines[get(L)]) < 0) {
            insertBefore(L, i);
            inserted = 1;
          }
        }
      }
      else { // If equal to cursor
        insertBefore(L, i);
        inserted = 1;
      }
    }
  }

  moveFront(L);
  while (index(L) != -1) {
    fprintf(out, "%s", lines[get(L)]);
    moveNext(L);
  }

  for (int i = 0; i < lineCount; i++) free(lines[i]); // frees each line
  free(lines); // frees whole array

  freeList(&L);
  fclose(in);
  fclose(out);
}
